File: voto.py
import os
import logging
import time
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from dotenv import load_dotenv

load_dotenv()

LINK = os.getenv('LINK')
PROXY_HOST = os.getenv('PROXY_HOST')
USE_PROXY = os.getenv('USE_PROXY') == 'true'
PROXY_PORT = os.getenv('PROXY_PORT')
PROXY_USERNAME = os.getenv('PROXY_USERNAME')
PROXY_PASS = os.getenv('PROXY_PASS')
INFO_TO_SEARCH = os.getenv('INFO_TO_SEARCH')
RADIO_BUTTON = os.getenv('RADIO_BUTTON')
VOTE_BUTTON = os.getenv('VOTE_BUTTON')

logging.basicConfig(level=logging.INFO)

def close_cookie_banner(driver):
    try:
        cookie_banner = WebDriverWait(driver, 30).until(
            EC.visibility_of_element_located((By.CSS_SELECTOR, '.styles__CookieContainer-sc-150pbzj-0'))
        )
        cookie_banner.click()
        logging.info('Banner de cookies fechado.')
    except:
        logging.warning('Banner de cookies não encontrado.')

def close_popups(driver):
    logging.info('Fechando popups...')
    try:
        notification_cancel_button = WebDriverWait(driver, 30).until(
            EC.visibility_of_element_located((By.XPATH, '//*[@id="onesignal-slidedown-cancel-button"]'))
        )
        notification_cancel_button.click()
        logging.info('Popup de notificação fechado.')
    except:
        logging.warning('Popup de notificação não encontrado.')

def clear_cache(driver):
    logging.info('Limpando cache...')
    driver.execute_script('window.localStorage.clear();')
    driver.execute_script('window.sessionStorage.clear();')
    driver.delete_all_cookies()
    logging.info('Cache limpo.')

def get_votes_for_item(driver):
    try:
        items = driver.find_elements(By.CSS_SELECTOR, '.pds-feedback-group')

        for item in items:
            title_element = item.find_element(By.CSS_SELECTOR, '.pds-answer-text')
            title = title_element.get_attribute('title')

            if title == INFO_TO_SEARCH:
                votes_element = item.find_element(By.CSS_SELECTOR, '.pds-feedback-votes')
                votes_text = votes_element.text
                logging.info(f'Votos para {INFO_TO_SEARCH}: {votes_text}')
                return votes_text
        logging.warning(f'{INFO_TO_SEARCH} não encontrado.')
    except Exception as error:
        logging.error(f'Erro ao buscar votos para {INFO_TO_SEARCH}: {error}')

def vote_multiple_times(site_url, n_votes):
    logging.info('Configurando driver...')
    wait5 = 5
    chrome_options = Options()

    if USE_PROXY:
        logging.info('Configurando proxy...')
        chrome_options.add_argument(f'--proxy-server=http://{PROXY_USERNAME}:{PROXY_PASS}@{PROXY_HOST}:{PROXY_PORT}')

    service = Service('chromedriver.exe')
    service.start()

    driver = webdriver.Chrome(service=service, options=chrome_options)

    for i in range(n_votes):
        logging.info(f'Votando {i + 1} de {n_votes}...')
        driver.get(site_url)
        close_popups(driver)
        close_cookie_banner(driver)

        try:
            radio_button = WebDriverWait(driver, 30).until(
                EC.visibility_of_element_located((By.XPATH, RADIO_BUTTON))
            )
            driver.execute_script('arguments[0].scrollIntoView(true);', radio_button)
            time.sleep(wait5)
            driver.execute_script('arguments[0].click();', radio_button);
            logging.info('Botão de escolha clicado.')
        except Exception as error:
            logging.error('O botão de escolha não é clicável:', error)

        try:
            vote_button = WebDriverWait(driver, 30).until(
                EC.visibility_of_element_located((By.XPATH, VOTE_BUTTON))
            )
            vote_button.click()
            logging.info('Botão de voto clicado.')
            time.sleep(wait5)
            get_votes_for_item(driver)
            time.sleep(wait5)
            clear_cache(driver)
            logging.info('Página atualizada.')
        except Exception as error:
            logging.error('O botão de voto não é clicável:', error)

    driver.quit()
    service.stop()
    logging.info('Finalizado.')

if __name__ == '__main__':
    import sys
    args = sys.argv[1:]
    n_votes = 10
    if '--votos' in args:
        n_votes = int(args[args.index('--votos') + 1])
    vote_multiple_times(LINK, n_votes)

File: voto.js
const { Builder, By, until } = require('selenium-webdriver');
const { Options } = require('selenium-webdriver/chrome');
const dotenv = require('dotenv');
const log = require('loglevel');
const process = require('process');

dotenv.config();

const LINK = process.env.LINK;
const PROXY_HOST = process.env.PROXY_HOST;
const USE_PROXY = process.env.USE_PROXY === 'true';
const PROXY_PORT = process.env.PROXY_PORT;
const PROXY_USERNAME = process.env.PROXY_USERNAME;
const PROXY_PASS = process.env.PROXY_PASS;
const INFO_TO_SEARCH = process.env.INFO_TO_SEARCH;
const RADIO_BUTTON = process.env.RADIO_BUTTON;
const VOTE_BUTTON = process.env.VOTE_BUTTON;

log.setLevel('info');

async function closeCookieBanner(driver) {
  try {
    const cookieBanner = await driver.wait(
      until.elementLocated(By.css('.styles__CookieContainer-sc-150pbzj-0')),
      30000
    );
    await driver.wait(until.elementIsVisible(cookieBanner), 30000);
    await cookieBanner.click();
    log.info('Banner de cookies fechado.');
  } catch {
    log.warn('Banner de cookies não encontrado.');
  }
}

async function closePopups(driver) {
  log.info('Fechando popups...');
  try {
    const notificationCancelButton = await driver.wait(
      until.elementLocated(
        By.xpath('//*[@id="onesignal-slidedown-cancel-button"]')
      ),
      30000
    );
    await driver.wait(until.elementIsVisible(notificationCancelButton), 30000);
    await notificationCancelButton.click();
    log.info('Popup de notificação fechado.');
  } catch {
    log.warn('Popup de notificação não encontrado.');
  }
}

async function clearCache(driver) {
  log.info('Limpando cache...');
  await driver.executeScript('window.localStorage.clear();');
  await driver.executeScript('window.sessionStorage.clear();');
  await driver.manage().deleteAllCookies();
  log.info('Cache limpo.');
}

async function getVotesForItem(driver) {
  try {
    const items = await driver.findElements(By.css('.pds-feedback-group'));

    for (let item of items) {
      const titleElement = await item.findElement(By.css('.pds-answer-text'));
      const title = await titleElement.getAttribute('title');

      if (title === INFO_TO_SEARCH) {
        const votesElement = await item.findElement(
          By.css('.pds-feedback-votes')
        );
        const votesText = await votesElement.getText();
        log.info(`Votos para ${INFO_TO_SEARCH}: ${votesText}`);
        return votesText;
      }
    }
    log.warn(`${INFO_TO_SEARCH} não encontrado.`);
  } catch (error) {
    log.error(`Erro ao buscar votos para ${INFO_TO_SEARCH}:`, error);
  }
}

async function voteMultipleTimes(siteUrl, nVotes) {
  log.info('Configurando driver...');
  const wait5 = 5000;
  const wait15 = 15000;
  const chromeOptions = new Options();

  if (USE_PROXY) {
    log.info('Configurando proxy...');
    chromeOptions.addArguments(
      `--proxy-server=http://${PROXY_USERNAME}:${PROXY_PASS}@${PROXY_HOST}:${PROXY_PORT}`
    );
  }

  const driver = new Builder()
    .forBrowser('chrome')
    .setChromeOptions(chromeOptions)
    .build();

  for (let i = 0; i < nVotes; i++) {
    log.info(`Votando ${i + 1} de ${nVotes}...`);
    await driver.get(siteUrl);
    await closePopups(driver);
    await closeCookieBanner(driver);

    try {
      const radioButton = await driver.wait(
        until.elementLocated(By.xpath(RADIO_BUTTON)),
        30000
      );
      await driver.wait(until.elementIsVisible(radioButton), 30000);
      await driver.executeScript(
        'arguments[0].scrollIntoView(true);',
        radioButton
      );
      await driver.sleep(wait5);
      await driver.executeScript('arguments[0].click();', radioButton);
      log.info('Botão de escolha clicado.');
    } catch (error) {
      log.error('O botão de escolha não é clicável:', error);
    }

    try {
      //   await driver.sleep(wait5);
      const voteButton = await driver.wait(
        until.elementLocated(By.xpath(VOTE_BUTTON)),
        30000
      );
      await driver.wait(until.elementIsVisible(voteButton), 30000);
      await voteButton.click();
      log.info('Botão de voto clicado.');
      await driver.sleep(wait5);
      await getVotesForItem(driver);
      await driver.sleep(wait5);
      await clearCache(driver);
      log.info('Página atualizada.');
    } catch (error) {
      log.error('O botão de voto não é clicável:', error);
    }
  }
  await driver.quit();
  log.info('Finalizado.');
}

const args = process.argv.slice(2);
const nVotes = args.includes('--votos')
  ? parseInt(args[args.indexOf('--votos') + 1])
  : 10;
voteMultipleTimes(LINK, nVotes);
